pkg/distributor/distributor.go

```diff
@@ -1305,8 +1305,14 @@ func (d *Distributor) push(ctx context.Context, pushReq *Request) error {
 	// Get a subring if tenant has shuffle shard size configured.
 	subRing := d.ingestersRing.ShuffleShard(userID, d.limits.IngestionTenantShardSize(userID))
 
-	// Use an independent context to make sure all ingesters get samples even if we return early
-	localCtx, cancel := context.WithTimeout(context.WithoutCancel(ctx), d.cfg.RemoteTimeout)
+	// Use an independent context to make sure all ingesters get samples even if we return early.
+	// It will still take a while to calculate which ingester gets which series,
+	// so we'll start the remote timeout once the first callback is called.
+	localCtx := context.WithoutCancel(ctx)
+	var cancelPushContext context.CancelFunc = func() {}
+	startRemoteTimeout := sync.OnceFunc(func() {
+		localCtx, cancelPushContext = context.WithTimeout(localCtx, d.cfg.RemoteTimeout)
+	})
```

Comment:
isn't there a race between swapping out the context in the first invocation and using a context in the next?

Reply:
maybe if you use `sync.OnceValue()` you can get around that

Reply:
Maybe split `ring.DoBatchWithOptions` into the slow init part and the make-calls part, so you don't need to change the variable in the callback?

Reply:
> isn't there a race between swapping out the context in the first invocation and using a context in the next?

All calls will wait for first func to finish. See the comment on `Once.Do`:
```
func (o *Once) Do(f func()) {
	// Note: Here is an incorrect implementation of Do:
	//
	//	if atomic.CompareAndSwapUint32(&o.done, 0, 1) {
	//		f()
	//	}
	//
	// Do guarantees that when it returns, f has finished.
```

Reply:
I've refactored using `sync.OnceValues` to make it more obvious.

I had to move the fast slab pool to the parent context, which is passed to `ring.DoBatch`, but that shouldn't be an issue (the key is conserved, and I understand that it doesn't depend on the context cancellation, it only sets itself on the context).

I would love to refactor this entire method further, extracting smaller pieces of code into separate functions, and reordering, but I'd rather do that in a separate PR.

Reply:
Follow up PR with the refactor: https://github.com/grafana/mimir/pull/6978/files